### ISSUE TITLE
Unable to load libgnutls-30.dll on a Azure Function deployed on a Windows plan on Azure

### ISSUE THREAD

**Issue by rob-hendrickx-c4t** [robinrodricks/FluentFTP.GnuTLS#76]
Whenever I try to use FluentFTP with GnuTLS enabled locally in an Azure Function, everything works as expected.

When deploying the function app to Azure on a Windows plan however, I bump into the following error:
> FluentFTP.GnuTLS.GnuTlsException: Could not load libgnutls-30.dll, (126) The specified module could not be found.

Even though the libgnutls-30.dll file is located in the same folder as the dll of my function app. I don't know how the DLL loading logic of GnuTLS works, but I think I might have encountered a similar issue in the past already.

Azure Functions running on Azure seem to be running from a different location from where the DLL is in fact located. This lead to the problem that e.g. Environment.CurrentDirectory points to a location like `D:\Program Files (x86)\SiteExtensions\Functions\2.0.12493\64bit` instead of `C:\home\site\wwwroot\bin` where my own binaries would be stored.

Therefor I am wondering if it is possible to add support to do either of the following:
- make it possible to define a custom location from which the FunctionLoader will search for the DLLs on Windows
- automatically do it relatively from the calling assembly's location?
- other solutions

**Comment by rob-hendrickx-c4t** [comment]
According to me it would help tremendously if an optional setting would be added that allows a path to be set where the DLLs are loaded from

When invoking the LoadLibrary, a Path.Combine on the setting value + the DLL filename would then replace the current argument (which is just the dll filename)

**Comment by rob-hendrickx-c4t** [comment]
@FanDjango that sounds fair 😊

**Comment by rob-hendrickx-c4t** [comment]
@FanDjango Just for our knowledge, do you happen to have an idea of when it could be fixed possibly? No pressure, and thanks already for the help! (and the help so far)

**Comment by FanDjango** [comment]
I'll have a test branch for you to play with later today.

**Comment by FanDjango** [comment]
Check out this branch I made:

https://github.com/FanDjango/FluentFTP.GnuTLS/tree/pathtest

and use additional config parameter:

```cs
		/// <summary>
		/// Add an optional string prefix to the LoadLibrary dllname
		/// </summary>

		public string LoadLibraryDllNamePrefix = string.Empty;
```

**Comment by rob-hendrickx-c4t** [comment]
@FanDjango Unfortunately this does not work it seems. 

> GnuTLS .dll load/call validation error
  ----> FluentFTP.GnuTLS.GnuTlsException : Could not load C:/temp/gnutls/libgnutls-30.dll, (126) The specified module could not be found.

If the dlls are not in the same folder as the executing assembly, I get this error. Even though the needed DLLs are in the location mentioned in the exception above

**Comment by FanDjango** [comment]
There seem to be a number of information sources about the dllName parameter for LoadLibrary, it's settings and also the search order for dll's and how to influence it.

It is even not totally clear (at this moment) to me, relative path only, or must it be an absolute path? In C#, of course you would use `@"....."` and single backslashes, but beware when simply quoting, then you need escaped backslashed. Oh, and since we call LoadLibrary **also** via PInvoke, maybe there is still a problem with the charset on the marshall for the dllName, thus losing the backslashes somehow? The exception message is constructed by us, it is not from the return of LoadLibrary itself.

**Comment by FanDjango** [comment]
An alternative to at least get some more information: **Where do you need to put the .dlls** _with no special prefix_, in order for them to be found? Have you identified a place where they should be put?

**Comment by rob-hendrickx-c4t** [comment]
@FanDjango Apologies, it was in fact related to the way I was testing it. 
I now repeated with a simpler test by temporarily adjusting locally the FluentFTP.LocalTests > GnuTests test suite and it works perfectly (with an absolute path, which is the approach I need))

**Comment by FanDjango** [comment]
In that case, I will speedily move towards merging and releasing this.

**Comment by FanDjango** [comment]
Not released yet. Please stand by.

**Comment by FanDjango** [comment]
Released 1.0.15, Feedback requested by all.

**Comment by c0h1b4** [comment]
I'm still getting this error:

```
Category: Microsoft.AspNetCore.Server.IIS.Core.IISHttpServer
EventId: 2
SpanId: 4e8735c253b55f28
TraceId: d2ff17d72d96ecc8ce672574600b0083
ParentId: 0000000000000000
RequestId: 8000023d-0000-f100-b63f-84710c7967bb

RequestPath: /Ftp/list

Connection ID "17365880164751245875", Request ID "8000023d-0000-f100-b63f-84710c7967bb": An unhandled exception was thrown by the application.

Exception: 

FluentFTP.GnuTLS.GnuTlsException: GnuTLS .dll load/call validation error

---> FluentFTP.GnuTLS.GnuTlsException: Could not load libgnutls-30.dll, (126) The specified module could not be found.

   --- End of inner exception stack trace ---

   at FluentFTP.GnuTLS.GnuTlsInternalStream.Validate(Boolean log)

   at FluentFTP.GnuTLS.GnuTlsInternalStream..ctor(String targetHostString, Socket socketDescriptor, CustomRemoteCertificateValidationCallback customRemoteCertificateValidation, X509CertificateCollection clientCertificates, String alpnString, GnuTlsInternalStream streamToResumeFrom, String priorityString, Boolean deInitGnuTls, Int32 handshakeTimeout, Int32 pollTimeout, GnuStreamLogCBFunc elog, Int32 logMaxLevel, GnuMessage logDebugInformationMessages, Int32 logQueueMaxSize)
   at FluentFTP.GnuTLS.GnuTlsStream.Init(BaseFtpClient client, String targetHost, Socket socket, CustomRemoteCertificateValidationCallback customRemoteCertificateValidation, Boolean isControl, IFtpStream controlConnStream, IFtpStreamConfig untypedConfig)
   at FluentFTP.FtpSocketStream.CreateCustomStream(String targetHost)
   at FluentFTP.FtpSocketStream.ActivateEncryption(String targetHost, X509CertificateCollection clientCerts, SslProtocols sslProtocols)
   at FluentFTP.FtpClient.Connect(Boolean reConnect)
   at FluentFTP.FtpClient.FluentFTP.IInternalFtpClient.ConnectInternal(Boolean reConnect)
   at FluentFTP.Client.BaseClient.BaseFtpClient.FluentFTP.IInternalFtpClient.ExecuteInternal(String command)
   at FluentFTP.Client.BaseClient.BaseFtpClient.ReadCurrentWorkingDirectory()
   at FluentFTP.Client.BaseClient.BaseFtpClient.FluentFTP.IInternalFtpClient.GetWorkingDirectoryInternal()
   at FluentFTP.FtpClient.GetWorkingDirectory()
   at FluentFTP.FtpClient.GetAbsolutePath(String path)
   at FluentFTP.FtpClient.GetListing(String path, FtpListOption options)
   at FluentFTP.FtpClient.GetListing(String path)
   at webapi.Services.UserService.getDiretorio(String matricula) in C:\MyProjects\FTPS\webapi\Services\UserService.cs:line 144

   at webapi.Services.UserService.Authenticate(AuthenticateRequest model) in C:\MyProjects\FTPS\webapi\Services\UserService.cs:line 44
   at webapi.Controllers.UsuariosController.Authenticate(AuthenticateRequest model) in C:\MyProjects\FTPS\webapi\Controllers\UserController.cs:line 25
   at lambda_method3(Closure, Object, Object[])
   at Microsoft.AspNetCore.Mvc.Infrastructure.ActionMethodExecutor.SyncActionResultExecutor.Execute(ActionContext actionContext, IActionResultTypeMapper mapper, ObjectMethodExecutor executor, Object controller, Object[] arguments)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.InvokeActionMethodAsync()
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.InvokeNextActionFilterAsync()

--- End of stack trace from previous location ---

   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Rethrow(ActionExecutedContextSealed context)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.InvokeInnerFilterAsync()

--- End of stack trace from previous location ---

   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeFilterPipelineAsync>g__Awaited|20_0(ResourceInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeAsync>g__Awaited|17_0(ResourceInvoker invoker, Task task, IDisposable scope)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeAsync>g__Awaited|17_0(ResourceInvoker invoker, Task task, IDisposable scope)
   at Microsoft.AspNetCore.Routing.EndpointMiddleware.<Invoke>g__AwaitRequestTask|6_0(Endpoint endpoint, Task requestTask, ILogger logger)
   at Microsoft.AspNetCore.Authorization.AuthorizationMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Authentication.AuthenticationMiddleware.Invoke(HttpContext context)
   at webapi.Helpers.JwtMiddleware.Invoke(HttpContext context, IUserService userService) in C:\MyProjects\FTPS\webapi\Helpers\JwtMiddleware.cs:line 31

   at Microsoft.AspNetCore.Server.IIS.Core.IISHttpContextOfT`1.ProcessRequestAsync()

```

**Comment by FanDjango** [comment]
Did you set the config parameter?

**Comment by c0h1b4** [comment]
Sorry, what config parameter?

**Comment by c0h1b4** [comment]
<img width="707" alt="Screenshot 2023-06-05 at 16 09 00" src="https://github.com/robinrodricks/FluentFTP.GnuTLS/assets/253084/510290aa-a872-4c8b-b80c-471eda6e4f1a">

How to add the LoadLibraryDllNamePrefix parameter?

**Comment by c0h1b4** [comment]
The lib is just an example. Both ways are not working. They give compile error. The definition is not found.

**Comment by FanDjango** [comment]
Sure you got 1.0.15?

**Comment by c0h1b4** [comment]
<img width="1022" alt="Screenshot 2023-06-05 at 16 18 05" src="https://github.com/robinrodricks/FluentFTP.GnuTLS/assets/253084/6588f7b3-5c47-4af8-a5f8-5142978e062c">
Yep

**Comment by FanDjango** [comment]
Okidoki, I will check and re-release

**Comment by FanDjango** [comment]
Re-released: 1.0.16.

**Comment by FanDjango** [comment]
```cs
			await Task.WhenAll(Enumerable.Repeat(1, 5).Select(async _ => {

				using (var client = new AsyncFtpClient("127.0.0.1", "blaap", "bloop")) {
					client.Config.EncryptionMode = FtpEncryptionMode.Explicit;
					client.Config.DataConnectionType = FtpDataConnectionType.PASV;
					client.Config.DownloadDataType = FtpDataType.Binary;
					client.Config.ValidateAnyCertificate = true;

					client.LegacyLogger = FTPLogEvent;

					client.Config.CustomStream = typeof(GnuTlsStream);
					client.Config.CustomStreamConfig = new GnuConfig() {
						//SecuritySuite = GnuSuite.Normal,
						//SecurityOptions = new List<GnuOption> {
						//		new GnuOption(GnuOperator.Include, GnuCommand.Protocol_All),
						//		new GnuOption(GnuOperator.Exclude, GnuCommand.Protocol_Tls10),
						//		new GnuOption(GnuOperator.Exclude, GnuCommand.Protocol_Tls11),
						//	},
						//SecurityProfile = GnuProfile.None,
						//AdvancedOptions = new List<GnuAdvanced> {
						//		GnuAdvanced.NoExtensions,
						//	},
						SetALPNControlConnection = string.Empty,
						SetALPNDataConnection = string.Empty,
						DeInitGnuTls = false,
						LoadLibraryDllNamePrefix = string.Empty,
						LogLevel = 2,
						LogMessages = GnuMessage.InteropFunction,
					};

					client.LegacyLogger = FTPLogEvent;

					await client.Connect();

					var item = await client.GetListing("", FtpListOption.ForceList);

					await client.Disconnect();
				}
			}
			));
```

**Comment by FanDjango** [comment]
It is working... in case your's does not, we need to open a new issue and, you know, logs, documentation etc. instead of hijacking this issue here.

**Comment by FanDjango** [comment]
Re-released 1.0.16 for safeties sake.
Closing the issue - feel free to comment if you want it to be re-opened.

**Comment by rob-hendrickx-c4t** [comment]
Unfortunately it seems that something is still off...

I now deployed an Azure Function app that consumes the library to Azure, but again get the error 
> GnuTLS .dll load/call validation error Could not load C:\home\site\wwwroot\bin\libgnutls-30.dll, (126) The specified module could not be found.

If I inspect the deployed app however, you can see that the DLL is located in the path described by the error. 

![image](https://github.com/robinrodricks/FluentFTP.GnuTLS/assets/38980805/7a0f8e11-6598-4fad-b98e-76b761902462)

Any idea what could be wrong?

**Comment by FanDjango** [comment]
File permissions, maybe? The user this is running under needs execution access to these files? Clutching at straws here...

One thing that you could try is to code **your own** test somewhere in your main program:

```cs
			private const string dllNameWinUtil = @"Kernel32.dll";
			private static extern IntPtr LoadLibrary([MarshalAs(UnmanagedType.LPStr)] string lpFileName);
			[DllImport(dllNameWinUtil, CallingConvention = CallingConvention.StdCall, CharSet = CharSet.Ansi)]
```

and 
```cs
string dllPath = "C:\\home\\site\\wwwroot\\bin\\libgnutls-30.dll";

					_ = SetErrorMode(ErrorModes.SEM_FAILCRITICALERRORS | ErrorModes.SEM_NOGPFAULTERRORBOX | ErrorModes.SEM_NOOPENFILEERRORBOX);
					_ = GetLastError();
					hModule = LoadLibrary(dllPath);

					if (hModule == IntPtr.Zero) {
						uint err = GetLastError();
						if (err != 0) {
							_ = FormatMessage(0x00001300, IntPtr.Zero, err, 0, ref errMsgPtr, 256, IntPtr.Zero);
						}
						if (errMsgPtr != IntPtr.Zero) {
							errMsg = Marshal.PtrToStringAnsi(errMsgPtr).TrimEnd(Environment.NewLine.ToCharArray());
						}
						throw new GnuTlsException("Could not load " + dllPath + ", (" + err + ") " + errMsg);
					}

```

and play around with that.

**Comment by rob-hendrickx-c4t** [comment]
@c0h1b4 I doubt that that's the issue in my case, since I'm using absolute paths in my logic (e.g. `C:\home\site\wwwroot\bin\`). 
Unfortunately, I didn't yet find a root cause for my issue. One thing that could be related is the usage of RunFromPackage in our case, where the folder I described above is marked "readonly" https://github.com/Azure/app-service-announcements/issues/84
Alternative methods to deploy/host the app are not yet successful though unfortunately

**Comment by FanDjango** [comment]
"mounted as a read-only Filesystem" - shouldn't hurt if execute privilege is honored

**Comment by rob-hendrickx-c4t** [comment]
I thought as much indeed, but that's the thing I'm not sure about. Trying to figure out if/how to ensure this is the case

**Comment by c0h1b4** [comment]
IMHO, it isn't a permission issue. If it were, the error would be about not having permission. It shows that the file is not where the program thinks it was supposed to be, hence 'the module could not be found'.

**Comment by rob-hendrickx-c4t** [comment]
In some Microsoft docs I think I saw that missing file permissions could also lead to this error. On top of that, I got a 404 Not Found when trying to delete a file from this readonly mount in my case.

Tomorrow morning I'm going to push some test code onto my app to do some queries using the File APIs against the dlls, in an attempt to narrow down the issue

**Comment by FanDjango** [comment]
@rob-hendrickx-c4t 

You perhaps can make use of this snippet, I use it sometimes:

Note: `PCFileName` contains the complete path to the file and the filename.

```cs
Log("Access to: " + PCFileName);
FileSecurity fSecurity = File.GetAccessControl(PCFileName);
AuthorizationRuleCollection accessRules = fSecurity.GetAccessRules(true, true, typeof(System.Security.Principal.NTAccount));
foreach (FileSystemAccessRule accessRule in accessRules) {
	Log(accessRule.IdentityReference.ToString() + " " + accessRule.AccessControlType.ToString() + " " + accessRule.FileSystemRights.ToString());
}
```

**Comment by rob-hendrickx-c4t** [comment]
Update
I can confirm that the libgnutls-30.dll file cannot be found in Environment.CurrentDirectory, but can be found in the location I specified in the prefix config setting now. 

Regarding file permissions, I see the following:
```
 [
    {
      "fileSystemRights": 2032127,
      "accessControlType": 0,
      "identityReference": {
        "value": "NT AUTHORITY\\SYSTEM"
      },
      "isInherited": false,
      "inheritanceFlags": 0,
      "propagationFlags": 0
    },
    {
      "fileSystemRights": 2032127,
      "accessControlType": 0,
      "identityReference": {
        "value": "BUILTIN\\Administrators"
      },
      "isInherited": false,
      "inheritanceFlags": 0,
      "propagationFlags": 0
    },
    {
      "fileSystemRights": 1245631,
      "accessControlType": 0,
      "identityReference": {
        "value": "IIS APPPOOL\\xxx
      },
      "isInherited": false,
      "inheritanceFlags": 0,
      "propagationFlags": 0
    }
```

The last one, which I assume my app runs under, has the "Modify" & "Synchronize" rights (https://learn.microsoft.com/en-us/dotnet/api/system.security.accesscontrol.filesystemrights?view=net-7.0), which should mean that execute is also granted to it. Except if the LoadLibrary operation runs under a different identity than this IIS user? But that I don't know for sure. 

Bottom line: it looks like file permissions are in fact correctly set for the dlls, so not sure what the issue is unfortunately.

**Comment by FanDjango** [comment]
@rob-hendrickx-c4t  Have you investigated this: https://github.com/robinrodricks/FluentFTP.GnuTLS/issues/76#issuecomment-1578699748

**Comment by FanDjango** [comment]
Note: 

126 (The specified module could not be found.) **also is reported, when** _one of the dependancies_ of the requested dll could not be **loaded** (or found).

You might want to verify that all dependancies are loadable at the location where you have the dll's by using a dependancy examination tool. I use https://github.com/lucasg/Dependencies/releases/tag/v1.11.1

**Comment by rob-hendrickx-c4t** [comment]
Intermediate update:
I can verify that loading libgmp-10.dll, libhogweed-6.dll, libnettle-8.dll, libwinpthread-1.dll & libgcc_s_seh-1.dll seem to be working correctly without error 126.

Using the tool you shared, that leaves following dependencies from system32 (which I would _**assume**_ to be working, but most likely one of them won't.

![image](https://github.com/robinrodricks/FluentFTP.GnuTLS/assets/38980805/156d11b7-f7b2-4d77-b082-b528021456a8)


I'll report back when I find more about the root cause.

**Comment by FanDjango** [comment]
The "Windows" ones will be found, because they are in the PATH. So the riddle continues.

Even if in the end, you won't want to solve the problem by putting your dll's into one of the directories in the PATH, for a first verification you could try that at least - in that case don't forget to **not** specify the prefix.

**Comment by rob-hendrickx-c4t** [comment]
@FanDjango Unfortunately I don't have access to any folder present in the PATH environment variable, nor do I have the permission to update that environment variable...

So I'm kind of lost in how I can proceed with my investigation unfortunately, any ideas?

To recap:
* All lib* dlls are found using basic I/O
* All these lib* dlls seem to have the execute right (through the "Modify" right)
* Using kernel32's LoadLibrary, I only get an error for libgnutls-30.dll, not for any of the other dlls included in the project.

**Comment by FanDjango** [comment]
That last one:

> Using kernel32's LoadLibrary, I only get an error for libgnutls-30.dll, not for any of the other dlls included in the project.

You see, when you load `libgnutls-30.dll`, this will scan for the presence of the dependencies. If they are not loaded, then they **too will be loaded**. Internally this is done in the same way, by whatever LoadLibrary calls internally, one could google this.

Unfortunately, I would now surmise that these **subsequent** loads will **fail**, as they won't have the benefit of the prefix. They work when you do them individually **with the prefix**, though.

I suppose you might succeed, if you successfully `LoadLibrary(...)` the dependant ones (only those that come with the FluentFTP.GnuTLS package) **first** and then finally try to load `libgnutls-30.dll'.

**Comment by rob-hendrickx-c4t** [comment]
I will try using LoadLibrary in a loop for all needed DLLs. I do want to note though that some other lib* dlls also have dependencies on eachother (e.g. libhogweed-6.dll depends on libnettle-8.dll) and those load just fine.

**Comment by rob-hendrickx-c4t** [comment]
@FanDjango I'll be damned...
```
{
  "assemblyDirectory": "C:\\home\\site\\wwwroot\\bin",
  "filesWithResult": [
    {
      "filePath": "C:\\home\\site\\wwwroot\\bin\\libgcc_s_seh-1.dll",
      "resultMessage": "success"
    },
    {
      "filePath": "C:\\home\\site\\wwwroot\\bin\\libgmp-10.dll",
      "resultMessage": "success"
    },
    {
      "filePath": "C:\\home\\site\\wwwroot\\bin\\libnettle-8.dll",
      "resultMessage": "success"
    },
    {
      "filePath": "C:\\home\\site\\wwwroot\\bin\\libwinpthread-1.dll",
      "resultMessage": "success"
    },
    {
      "filePath": "C:\\home\\site\\wwwroot\\bin\\libhogweed-6.dll",
      "resultMessage": "success"
    },
    {
      "filePath": "C:\\home\\site\\wwwroot\\bin\\libgnutls-30.dll",
      "resultMessage": "success"
    }
  ]
}
```

For reference, this is a snipet of the code I behind this test


```
 [FunctionName("LoadSequence")]
public static async Task<IActionResult> RunLoadSequence(
               [HttpTrigger(AuthorizationLevel.Function, "post", Route = null)] HttpRequest req,
               ILogger log)
{
    var assemblyDirectory = Path.GetDirectoryName(typeof(Test).Assembly.Location);


    var filesWithResult = new List<Result>
    {
        new Result("libgcc_s_seh-1.dll"),
        new Result("libgmp-10.dll"),
        new Result("libnettle-8.dll"),
        new Result("libwinpthread-1.dll"),
        new Result("libhogweed-6.dll"),
        new Result("libgnutls-30.dll"),
    };

    foreach (var file in filesWithResult)
    {
        file.ResultMessage = TryLoadModule(file.FilePath);
    }

    return new OkObjectResult(new
    {
        assemblyDirectory,
        filesWithResult
    });
}

private static string TryLoadModule(string path)
{
    var errMsgPtr = IntPtr.Zero;
    var errMsg = string.Empty;

    _ = SetErrorMode(ErrorModes.SEM_FAILCRITICALERRORS | ErrorModes.SEM_NOGPFAULTERRORBOX | ErrorModes.SEM_NOOPENFILEERRORBOX);
    _ = GetLastError();
    var hModule = LoadLibrary(path);

    if (hModule == IntPtr.Zero)
    {
        var err = GetLastError();
        if (err != 0)
        {
            _ = FormatMessage(0x00001300, IntPtr.Zero, err, 0, ref errMsgPtr, 256, IntPtr.Zero);
        }
        if (errMsgPtr != IntPtr.Zero)
        {
            errMsg = Marshal.PtrToStringAnsi(errMsgPtr).TrimEnd(Environment.NewLine.ToCharArray());
        }
        return $"{err} ({errMsg})";
    }

    return "success";
}
```

Now, is this something I should code in my solution, or will it be made part of the FluentFTP.GnuTLS library? Either works for me, just to avoid we do it twice in our case

**Comment by FanDjango** [comment]
>  if someone would provide a directory as prefix (which in fact should always be empty or a directory, a fileName prefix won't work with the dependencies either), I assume this logic will always need to be applied to not run into the issue I faced. So not sure if an additional configuration toggle is needed aside from the prefix itself

YES! You are right. He who needs prefix, he also needs this sequential loading. Genius.

**Comment by FanDjango** [comment]
Checking...

**Comment by FanDjango** [comment]
released as 1.0.17 - the staggered loading.

Now we can address the next problem... sigh.

**Comment by FanDjango** [comment]
I just did a Nuget update to 1.0.17 on my test app and all is OK.

**Comment by FanDjango** [comment]
@rob-hendrickx-c4t Open the (-64) GNUTLS_E_FILE_ERROR fail as a new issue, please. That way people searching for problems will get a better pointer to it.

I prefer to close this one again.
